### DISCUSSION TITLE
Targeting ZCU111 with Vivado/Petalinux 2022.2

### DISCUSSION THREAD

**Discussion by drew-os** [srsran/zynq_timestamping#47] [General] [answered]
This is similar to https://github.com/srsran/zynq_timestamping/discussions/33 with the key difference being the version 2022.2 .

Background:
I'm a hardware engineer that is very new to the petalinux side of things.  I tried following the directions for the 2019.2 build, but ran into issues. Specifically at the petalinux-build stage there was an error along the lines of `fatal: cannot use gawk builtin 'namespace' as variable`, which is apparently a known issue with gawk 5.0+ and libgpg-error.  I wasn't sure what to do with that, so long story short I tried using 2022.2 tools and was able to actually build the petalinux project.

Issue:
I've successfully generated the bitstream and .xsa from Vivado 2022.2, created and built petalinux project, and compiled the customized linux kernel from source.  I'm at the point in the guide "Compiling a custom Linux kernel driver", and for the `make ARCH=arm64 -j4` command received the following error:

<pre>make ARCH=arm64 -j4
make -C /home/drew/Projects/linux-xlnx/ M=/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module modules
make[1]: Entering directory &apos;/home/drew/Projects/linux-xlnx&apos;
  CC [M]  /home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.o
<b>/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.c:</b> In function ‘<b>allocate_trx_dma_buffers</b>’:
<b>/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.c:424:22:</b> <font color="#CC0000"><b>error: </b></font>implicit declaration of function ‘<b>dma_zalloc_coherent</b>’; did you mean ‘<b>dma_alloc_coherent</b>’? [<font color="#CC0000"><b>-Werror=implicit-function-declaration</b></font>]
  424 |   buffer-&gt;virtaddr = <font color="#CC0000"><b>dma_zalloc_coherent</b></font>(&amp;d_info-&gt;pdev-&gt;dev, alloc_request-&gt;buffer_size,
      |                      <font color="#CC0000"><b>^~~~~~~~~~~~~~~~~~~</b></font>
      |                      <font color="#4E9A06">dma_alloc_coherent</font>
<b>/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.c:424:20:</b> <font color="#75507B"><b>warning: </b></font>assignment to ‘<b>void *</b>’ from ‘<b>int</b>’ makes pointer from integer without a cast [<font color="#75507B"><b>-Wint-conversion</b></font>]
  424 |   buffer-&gt;virtaddr <font color="#75507B"><b>=</b></font> dma_zalloc_coherent(&amp;d_info-&gt;pdev-&gt;dev, alloc_request-&gt;buffer_size,
      |                    <font color="#75507B"><b>^</b></font>
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:277: /home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.o] Error 1
make[1]: *** [Makefile:1868: /home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module] Error 2
make[1]: Leaving directory &apos;/home/drew/Projects/linux-xlnx&apos;
make: *** [Makefile:16: default] Error 2</pre>

The variable KDIR is set to `KDIR = /home/drew/Projects/linux-xlnx/` which is the location of the xilinx linux kernel compiled in the previous step.  Not really sure where to go from here and any help is appreciated!

Followup question:
If this should just be a separate discussion topic let me know, but would it be easier to add in the txrx_test, kernel module, device tree, and possibly even srsRAN source into the petalinux build process?  Something along the lines of `petalinux-create -t apps` and `petalinux-create -t modules`?  It seems to me this would remove the need to create the SDK, compile the kernel, modify the device tree, and compile the app and module outside of the petalinux tools environment.  I mean that could all be done within the petalinux tools so at the end you can create a BSP that neatly wraps it all together for future project builds.  At least that's my limited understanding of it.  Anyone familiar enough with that process to know what the pitfalls are?

**Comment by drew-os** [comment]
So the answer to the original problem was to replace dma_zalloc_coherent with dma_alloc_coherent exactly like the error description suggested.  Apparently dma_zalloc_coherent is deprecated and not present in new kernel versions.  If I run into any other issues I'll post them here.

**Reply by drew-os** [reply]
Ran into another issue, this time it's with the prepare.sh script and SIMD support.  Script output is
<pre><font color="#4E9A06"><b>drew@drew-VirtualBox</b></font>:<font color="#3465A4"><b>~/Projects/zynq_timestamping/app</b></font>$ ./prepare.sh zcu111
mkdir: cannot create directory ‘/home/drew/Projects/zynq_timestamping/app/bin_app’: File exists
--&gt; Current UE configuration
• Board IP: 10.12.1.201
• freq_offset: -4100
• rx_gain: 50
• time_adv_nsamples: 82

mkdir: cannot create directory ‘build_arm’: File exists
Preparing for cross-compilation
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /opt/petalinux/rfsoc/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- FFTW3F LIBRARIES: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libfftw3f.so
-- FFTW3F STATIC LIBRARIES: FFTW3F_STATIC_LIBRARY-NOTFOUND
-- FFTW3F INCLUDE DIRS: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
-- FFT_LIBRARIES: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libfftw3f.so
-- Checking for module &apos;mbedtls&apos;
--   No package &apos;mbedtls&apos; found
-- MBEDTLS LIBRARIES: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libmbedcrypto.so
-- MBEDTLS STATIC LIBRARIES: MBEDTLS_STATIC_LIBRARIES-NOTFOUND
-- MBEDTLS INCLUDE DIRS: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
-- PCSC LIBRARIES: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libpcsclite.so
-- PCSC INCLUDE DIRS: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include/PCSC
-- Building with PCSC support.
-- UHD LIBRARIES UHD_LIBRARIES-NOTFOUND
-- UHD INCLUDE DIRS UHD_INCLUDE_DIRS-NOTFOUND
-- Could NOT find UHD (missing: UHD_LIBRARIES UHD_INCLUDE_DIRS) 
-- SKIQ LIBRARIES SKIQ_LIBRARY-NOTFOUNDSKIQ_LIBRARY_GLIB-NOTFOUNDSKIQ_LIBRARY_USB-NOTFOUND/opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libtirpc.so.3
-- SKIQ INCLUDE DIRS SKIQ_INCLUDE_DIRS-NOTFOUND
-- Could NOT find SKIQ (missing: SKIQ_INCLUDE_DIRS) 
-- Checking for module &apos;libbladeRF&apos;
--   No package &apos;libbladeRF&apos; found
-- libbladeRF not found.
-- FINDING SOAPY.
-- Checking for module &apos;SoapySDR&apos;
--   No package &apos;SoapySDR&apos; found
-- libSOAPYSDR not found.
-- FINDING ZEROMQ.
-- Checking for module &apos;ZeroMQ&apos;
--   No package &apos;ZeroMQ&apos; found
-- libZEROMQ not found.
-- SRSGUI LIBRARIES SRSGUI_LIBRARIES-NOTFOUND
-- SRSGUI INCLUDE DIRS SRSGUI_INCLUDE_DIRS-NOTFOUND
-- Could NOT find SRSGUI (missing: SRSGUI_LIBRARIES SRSGUI_INCLUDE_DIRS) 
-- Could NOT find libdwarf (missing: LIBDWARF_LIBRARY LIBDWARF_INCLUDE_DIR) 
-- Building with backward-cpp support
-- Using standard LTE sampling rates
<font color="#CC0000">CMake Error at CMakeLists.txt:486 (message):</font>
<font color="#CC0000">  no SIMD instructions found</font>


-- Configuring incomplete, errors occurred!
See also &quot;/home/drew/Projects/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeOutput.log&quot;.
See also &quot;/home/drew/Projects/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeError.log&quot;.
make: *** No rule to make target &apos;usrp_txrx&apos;.  Stop.
Copying configuration files...
cp: cannot stat &apos;lib/examples/usrp_txrx&apos;: No such file or directory
cp: cannot stat &apos;lib/examples/pdsch_ue&apos;: No such file or directory
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /opt/petalinux/rfsoc/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- libmetal library found
--   LIBMETAL_INCLUDE_DIR: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
--   LIBMETAL_LIBRARIES: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libmetal.so
-- RFDC library found
--   RFDC_INCLUDE_DIR: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
--   RFDC_LIBRARY: /opt/petalinux/rfsoc/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/librfdc.so
-- /home/drew/Projects/zynq_timestamping/sw/build_arm/install/lib does not exist. Defaulting install location to /home/drew/Projects/zynq_timestamping/sw/build_arm/install/lib.
<font color="#CC0000">CMake Error at CMakeLists.txt:247 (message):</font>
<font color="#CC0000">  no SIMD instructions found</font>


-- Configuring incomplete, errors occurred!
See also &quot;/home/drew/Projects/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeOutput.log&quot;.
See also &quot;/home/drew/Projects/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeError.log&quot;.
make: *** No targets specified and no makefile found.  Stop.
cp: cannot stat &apos;./install/lib/*&apos;: No such file or directory
Done</pre>

I looked around in the petalinux config settings and didn't see anything about SIMD, SSE, or AVX.  From what I can gather these instructions are either supported by the CPU or not, and it should be handled by the compiler with the -march=arm64 option.  

Something that is different in v2022.2 is when installing the SDK, the environment setup script is named `/opt/plnx_sdk_rfsoc/environment-setup-cortexa72-cortexa53-xilinx-linux` instead of `/opt/plnx_sdk_rfsoc/environment-setup-aarch64-xilinx-linux`.  Maybe not an issue, but perhaps it doesn't contain an expected setup variable?

**Reply by sauka** [reply]
Hey @drew-os ,

Thanks for your feedback! You are right, the problem is that in v2022.2 the sysroot is built as  `*sysroots/cortexa72-cortexa53-xilinx-linux` and **toolchain.cmake** will set `CMAKE_SYSTEM_PROCESSOR` to **cortexa72** (or **cortexa72-cortexa53**), you can check it by adding to [CMakeLists.txt](https://github.com/srsran/srsRAN/blob/master/CMakeLists.txt) the following line:

`message(STATUS " CMAKE_SYSTEM_PROCESSOR is ${CMAKE_SYSTEM_PROCESSOR} ")`

I'd just modify the [condition](https://github.com/srsran/srsRAN/blob/master/CMakeLists.txt#L466) so that it treats your processor the same as aarch64 and the error should go away then

**Reply by drew-os** [reply]
Hi @sauka,

That worked!  Now I've run into an issue with running the txrx_test example:

```
xilinx-zcu111-20222:~/srs/bin_app$ export LD_LIBRARY_PATH=/home/petalinux/srs/bin_app/
xilinx-zcu111-20222:~/srs/bin_app$ alias sudo='sudo PATH="$PATH" HOME="$HOME" LD_LIBRARY_PATH="$LD_LIBRARY_PATH"'
xilinx-zcu111-20222:~/srs/bin_app$ sudo nice -20 ./txrx_test -f 2400000000 -a n_prb=6 -p 6 -g 40 -o test_txrx_zcu.bin
Opening RF device...
Active RF plugins: libsrsran_rf_rfdc.so
Inactive RF plugins:
Supported RF device list: RFdc
Trying to open RF device 'RFdc'
CH0 n_prb=6
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
Configuring LMK04208 to use internal clock source
LMX configured
metal: error:
 ADC 0 timed out at state 6 in XRFdc_WaitForRestartClr
/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/rf_xlnx_rfdc_imp.c:610: ERROR: Failed to wake up ADC tile 1
Unable to open RF device 'RFdc'
/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/rf_imp.c:152: Failed to open a RF frontend device. Please check the available srsRAN CMAKE options to verify what RF frontend devices have been detected in your system
/home/drew/Projects/zynq_timestamping/app/srsRAN/lib/examples/usrp_txrx.c:140: Error opening rf
```
I had to alias the sudo command in order to load the LD_LIBRARY_PATH variable in the sudo environment, FYI.  

I'm not sure why it's unable to communicate with the RF device, but it may be driver related.  I had to deviate from the guide's driver build method because there was a kernel version discrepancy that prevented installation of the driver: 
```
xilinx-zcu111-20222:~/srs$ insmod srs_dma_driver.ko
[   53.689005] srs_dma_driver: version magic '5.15.0 SMP mod_unload aarch64' should be '5.15.36-xilinx-v2022.2 SMP mod_unload aarch64'
insmod: ERROR: could not insert module srs_dma_driver.ko: Invalid module format
```
Instead of chasing down that error, I copied the driver source into a new petalinux module and compiled it there with the following commands:
```
petalinux-create -t modules -n srs-dma-driver --enable
cp ~/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/kernel_module/srs_dma_driver.c ~/Projects/petalinux_builds_2022.2/srs_timestamping/project-spec/meta-user/recipes-modules/srs-dma-driver/files/srs-dma-driver.c
petalinux-build -c srs-dma-driver
```
It built without error, so I then modified `~/Projects/petalinux_builds_2022.2/srs_timestamping/project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi` to include the device tree changes from the guide. Only the `srs_rx_dma {};` and `srs_tx_dma {};` sections though since the DMA parts get included by the import of the Vivado .XSA file.  I then re-built the whole petalinux project, packaged it, and put it on the SD card.

Any clue what could be going on?

**Comment by sauka** [comment]
Hi @drew-os ,

Thanks for the detailed description of driver compilation using Petalinux tools - this is another way to do it compared to manual compilation we described in the appnote. 
The initial error you got when you tried to load `srs_dma_driver.ko` was caused by Linux kernel version mismatch between the one compiled in Petalinux project and one you used following the appnote (probably you checked out wrong tag/branch?). Anyway, glad you solved it! We will use the info you shared here to extend the appnote :+1: 

`checking driver vfio-platform,a0000000.usp_rf_data_converter`
`metal: error:`

I'd say the fpga design and device-tree node for RFdc IP mismatch. Although I understood that you relied on Petalinux to build it (based on exported hardware). Basically I would first make sure that the address of RFdc IP in Vivado project and device-tree match. Then another thing that I encountered myself a few times - when you upgrade some IP (custom or Xilinx) due to a new version availability, sometimes the IP looses its configuration, thus please check the GUI config of the RFdc IP and compare it to the original 2019.2 project.


Small note regarding device-tree:
> It built without error, so I then modified ~/Projects/petalinux_builds_2022.2/srs_timestamping/project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi to include the device tree changes from the guide. Only the srs_rx_dma {}; and srs_tx_dma {}; sections though since the DMA parts get included by the import of the Vivado .XSA file. I then re-built the whole petalinux project, packaged it, and put it on the SD card.

Very good! Just make sure that the `srs_rx_dma` and `srs_tx_dma` nodes you copied do reference correct Xilinx DMA nodes in your device-tree.

**Reply by drew-os** [reply]
I'm glad that was helpful!
So I went back and looked at the addresses in Vivado and the device tree, and found a small discrepancy with the phandle referencing the DMA's
```
srs_rx_dma {
		compatible = "srs,txrx_dma";
		dmas = <&axi_dma_0 0x00>;
		dma-names = "rx";
		dma-direction = "rx";
		dma-coherent;
	};

	srs_tx_dma {
		compatible = "srs,txrx_dma";
		dmas = <&axi_dma_1 0x00>;
		dma-names = "tx";
		dma-direction = "tx";
	};
```
It's not a big change, just adding &axi_dma_0/1 to reference the labels from the `components/plnx_workspace/device-tree/device-tree/pl.dtsi` auto-generated file. 

After that I tried it again and it still wasn't working, so I tried running the rfdc-selftest app from xilinx and got this error
```
xilinx-zcu111-20222:~$ rfdc-selftest
RFdc Selftest Example Test
metal: warning:   metal_linux_irq_handling: Failed to set scheduler: Unknown error -1.
metal: info:      metal_linux_dev_open: checking driver vfio-platform,b0080000.usp_rf_data_converter,(null)
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:     metal_linux_exec_cmd: 843: open(/sys/bus/platform/devices/b0080000.usp_rf_data_converter/driver_override): errno=13
metal: error:
 Failed to open device b0080000.usp_rf_data_convertermetal: error:     metal_linux_dev_open: udev_device platform:  not found
metal: error:
 Failed to open device   Selftest Example Test failed
```
Seems pretty similar to the error I get with the SRS driver.  I created another project from the ZCU111 2022.2 BSP and used the images from the pre-built directory, and get the same error! So something is amiss with the 2022.2 build.  I'll contact xilinx support and try to hunt this down.

**Reply by drew-os** [reply]
That last one was my mistake.  I needed to run the commands with 'sudo'.  After that the rfdc-selftest and rfdc-read-write examples work correctly, including with the srs_dma_driver installed.  Below is the output after trying to run the SRS txrx_test. Same error as a couple posts above.
```
xilinx-zcu111-20222:~/bin_app$ alias sudo='sudo PATH="$PATH" HOME="$HOME" LD_LIBRARY_PATH="$LD_LIBRARY_PATH"'
xilinx-zcu111-20222:~/bin_app$ sudo ./txrx_test -f 2400000000 -a n_prb=6 -p 6 -g 40 -o test_txrx_zcu.bin
Opening RF device...
Active RF plugins: libsrsran_rf_rfdc.so
Inactive RF plugins:
Supported RF device list: RFdc
Trying to open RF device 'RFdc'
CH0 n_prb=6
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
metal: info:      metal_linux_dev_open: checking driver vfio-platform,a0000000.usp_rf_data_converter,(null)
Configuring LMK04208 to use internal clock source
LMX configured
metal: error:
 ADC 0 timed out at state 6 in XRFdc_WaitForRestartClr
/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/xrfdc/rf_xlnx_rfdc_imp.c:610: ERROR: Failed to wake up ADC tile 1
Unable to open RF device 'RFdc'
/home/drew/Projects/zynq_timestamping/sw/lib/src/phy/rf/rf_imp.c:152: Failed to open a RF frontend device. Please check the available srsRAN CMAKE options to verify what RF frontend devices have been detected in your system
/home/drew/Projects/zynq_timestamping/app/srsRAN/lib/examples/usrp_txrx.c:140: Error opening rf
xilinx-zcu111-20222:~/bin_app$ lsmod
Module                  Size  Used by
uio_dmem_genirq        16384  0
srs_dma_driver         20480  0
uio_pdrv_genirq        16384  0
```
I think this may be driver related.  Any thoughts?

**Comment by ofontbach** [comment]
Hi @drew-os, from #52 it seems that you've managed to overcome the issues detailed above. Is this correct? If that's so, would you mind to provide some minimal details and close this thread? Thanks!

**Reply by drew-os** [reply]
I never did get it to work with 2022.2.  Everything seemed to build without errors, but the test program couldn't utilize the RFDC driver.  That's as far as I got.  Instead I switched back to 2019.2 version of tools with Ubuntu 18.04 and that was able to build and run.

**Comment by icchalmers** [comment, marked as answer]
Just for anyone else that ends up here in the future, I've managed to migrate the zcu111 build to Vivado/Petalinux 2023.2. Thanks to @drew-os for doing most of the legwork!

The final hurdle at the `ADC 0 timed out at state 6 in XRFdc_WaitForRestartClr` error (which basically says the ADC PLL isn't locked) is that the i2c bus numbers seem to have moved around a bit in newer Petalinux builds. Trying to program the clock tree silently fails, and the clocks to the RFdc are never generated.

For me, the SCI18IS60C had moved from bus `12` to bus `11`, as verified by i2cdetect (`2f` is the address of the i2c to spi chip used to talk to the LMK/LMX chips):

```bash
zcu111:~# i2cdetect -y -r 11
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 2f 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- 48 -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- UU UU -- -- 
```

Updating the `I2CBUS` variable [here](https://github.com/srsran/zynq_timestamping/blob/38ba060b5a64387488534ae1bfc4b78d35342dbe/sw/lib/src/phy/rf/xrfdc/rf_xlnx_rfdc_imp.c#L45) from `12` to `11` was enough for me to be able to run the txrx example successfully. A more robust approach would be to scan for the correct i2c bus as suggested in [this post](https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/84541826/Programming+Clocks+on+the+ZCU111), but the quick-fix was enough for me.

A modification I haven't seen mentioned elsewhere: when doing the IP upgrade in Vivado it has a habit of clobbering some of the IP settings. For me, the main one was that it resets the reset polarity on `rfdc_adc_data_decim_0` and `rfdc_dac_data_interp_0`. I assume they incorrectly have the reset polarity set as negative during the IP packaging (possibly due to the `N` in the signal name, which I have seen Vivado use in the past to infer polarity). The quick-fix is to just fix the polarity before re-exporting `design_1.tcl`:

```tcl
  set_property -dict [ list \
   CONFIG.POLARITY {ACTIVE_HIGH} \
 ] [get_bd_pins /rfdc_adc_data_decim_0/ADCxN_reset]

  set_property -dict [ list \
   CONFIG.POLARITY {ACTIVE_HIGH} \
 ] [get_bd_pins /rfdc_dac_data_interp_0/DACxN_reset]
```

The final detail I had different from the above discussions was I had to fix the DMA reference in the device tree for the rx channel (from `0` to a `1`).  The kernel logs were showing the tx channel being successfully created, but I was getting `srs_dma_driver srs_rx_dma: Couldn't find DMA channel: rx` on the rx channel. The final entry in `system-user.dtsi` was:

```
/include/ "system-conf.dtsi"
/ {
};

&amba_pl{
    srs_rx_dma {
        compatible = "srs,txrx_dma";
        dmas = <&axi_dma_0 0x01>;
        dma-names = "rx";
        dma-direction = "rx";
        dma-coherent;
    };

    srs_tx_dma {
        compatible = "srs,txrx_dma";
        dmas = <&axi_dma_1 0x00>;
        dma-names = "tx";
        dma-direction = "tx";
    };
};
```
Cheers,

Iain

**Reply by ofontbach** [reply]
Thank you very much for sharing this, @icchalmers. It's really nice to know that the timestamping solution can be used with 2022.2 :)
